### ISSUE TITLE
Redesign bulk-transfer interface

### ISSUE THREAD

**Issue by benharsh** [chapel-lang/chapel#5310]
## Existing Bulk-Transfer Interface
Here is our existing interface:
```chpl
proc BaseArr.dsiSUpportsBulkTransfer() param : bool;
proc BaseArr.dsiSupportsBulkTransferInterface() param : bool;
proc BaseArr.doiCanBulkTransferStride(viewDom) param : bool;

proc BaseArr.doiCanBulkTransfer(viewDom) : bool;
proc BaseArr.doiUseBulkTransfer(B) : bool;
proc BaseArr.doiUseBulkTransferStride(B) : bool;

proc BaseArr.doiBulkTransferFrom(Barg, viewDom);
proc BaseArr.doiBulkTransferToDR(Barg, viewDom);
proc BaseArr.doiBulkTransferFromDR(Barg, viewDom);
proc BaseArr.doiBulkTransfer(B, viewDom);
proc BaseArr.doiBulkTransferStride(B, viewDom);
```
I think the existing bulk-transfer interface is unnecessarily complicated by the number of functions and separation between contiguous, strided, and distributed transfers. I think we can develop a simpler and more flexible interface.

## Proposed Interface
With this new interface my hope is that we can avoid embedding the following questions into the API:
- Does this transfer involve a DefaultRectangular?
- Is this a strided transfer
- Can/should I use a bulk transfer?

Here is the proposed interface:
```chpl
// srcClass and destClass inherit from BaseArr
proc srcClass.doiBulkTransferToKnown(srcDom:domain, destClass, destDom:domain) : bool {}
proc destClass.doiBulkTransferFromKnown(destDom:domain, srcClass, srcDom:domain) : bool {}

proc srcClass.doiBulkTransferToAny(srcDom:domain, destClass, destDom:domain) : bool {}
proc destClass.doiBulkTransferFromAny(destDom:domain, srcClass, srcDom:domain) : bool {}
```
The following guarantees are made:
- srcClass.rank == srcDom.rank
- destClass.rank == destDom.rank
- srcDom.size == destDom.size

Implementers should know that:
- it is valid for srcClass.rank != destClass.rank (for rank changes)
- it is valid for the domains to not be distributed, even if the classes implement distributed domains

I believe we need to distinguish between 'Known' and 'Any' if we want to be able to choose the 'best' bulk transfer available. Let's say we only had ``doiBulkTransferTo`` and ``doiBulkTransferFrom``, and we're attempting a transfer between BlockDist and a third-party domain map. Let's also say that the third-party domain map knows how to transfer **to/from** BlockDist **or** DefaultRectangular.

We want BlockDist to have a fallback for bulk-transfer if the other array can transfer to/from DefaultRectangular arrays. If we ask BlockDist if it can transfer first, it will say "yes" because it sees that the third-party domain map can transfer with DefaultRectangular arrays. But the best option would be to use the third-party domain map's transfer to/from BlockDist. Having separate 'Known' and 'Any' methods allows us to check for a 'best' transfer first before attempting a fallback like BlockDist's DefaultRectangular portability transfer.

Domain map authors should implement ``doiBulkTransfer*Known`` methods with type constraints and where-clauses for domain-maps they can handle the best. While most 'Known' methods will target a specific domain map, the situation is different for something like an ArrayView. ArrayViews want to just invoke a bulk transfer on their underlying arrays after adjusting the domain argument appropriately. Since ArrayViews will always do the same thing for any domain map, they will implement their 'Known' methods as fully generic so that they are always called. Let's consider a statement like this involving assignment between slices of DefaultRectangular arrays:
```chpl
A[1..5] = B[1..5];
```

The bulk-transfer will more or less follow this pattern:
```
bulkTransfer(A, {1..5}, B, {1..5});              -> sees A, an ArrayViewSlice, supports doiBulkTransferFromKnown
A.doiBulkTransferFromKnown({1..5}, B, {1..5});   -> forwards the transfer to its underlying array
bulkTransfer(A.arr, {1..5}, B, {1..5});          -> sees B supports doiBulkTransferToKnown
B.doiBulkTransferToKnown({1..5}, A.arr, {1..5}); -> forwards the transfer to its underlying array
bulkTransfer(A.arr, {1..5}, B.arr, {1..5});      -> now a bulk transfer between DefaultRectangular arrays
```
Rank-changes and reindexes will transform their domain arguments on the way 'down'.

Unlike the existing interface, I propose that the doiBulkTransfer\* methods return a ``bool`` indicating whether the bulk-transfer was successful. This replaces the need for a method like ``doiUseBulkTransfer``. Anyone manually calling a doiBulkTransfer\* method should check the result and complete the transfer if unsuccessful. Our ``proc =(A:[],B:[])`` implementation will fall back to element-by-element assignment.

The doiBulkTransfer\*Any methods are a last chance for domain maps to do something better than element-by-element transfer. I imagine many domain map authors will implement bulk-transfers with DefaultRectangular arrays. Knowing this, we will implement BlockDist's doiBulkTransfer\*Any methods to use the other domain map's support for DefaultRectangular array transfers (if such support exists).

## RankChange Concerns
Rank-changes can be tricky to implement, even for the simple DefaultRectangular array. Consider the following example:
```chpl
var A, B : [1..10, 1..10] int;
A[.., 1] = B[1, ..];
```
When we get down to the transfer between DefaultRectangular arrays (after going through ArrayViews), destDom will be ``{1..10, 1..1}`` and srcDom will be ``{1..1, 1..10}``. At this point we would need to check the sizes of each dimension to determine the original rank of the transfer and which dimensions were collapsed.

I propose that we require domain map authors to **opt-in** to bulk-transfers involving rank changes by implementing the ``doiCanBulkTransferRankChange`` param method that returns a ``bool``. This is a lame solution and exactly the kind of thing I want to get away from in our current interface, but rank changes are sufficiently tricky that I think it's worth protecting domain map authors by default.

## Conclusion
While this would be a large-ish change, I don't think we have many (any?) users writing their own distributions and implementing the bulk-transfer interface. I think the proposed options are sufficiently general to handle the desired cases, and are much simpler than the current interface.

Regardless of the chosen option, we should finally document the bulk-transfer interface somewhere.

We should also consider the fate of the various ``useBulkTransfer`` config params.

**Comment by benharsh** [comment]
@bradcray : you may be interested in this topic. I'm not sure if I should advertise it more widely until array-views are more fully fleshed-out.

**Comment by gbtitus** [comment]
I'm interested in this at least in passing.  I added some of the `doiUseBulkTransfer*()` execution-time testing in order to differentiate between multi-ddata arrays that actually do have multiple data blocks (where bulk transfer is a little more complicated) and multi-ddata arrays which only have one data block and which therefore work just like the single-ddata arrays.  It sounds like here that difference would be covered by the `bool` return value from the new `doiBulkTransfer*()` functions.

**Comment by mppf** [comment]
@benharsh - be careful of something like
```chpl
proc BaseArr.doiCanBulkTransferTo(ref Dest: [], const viewDom: domain) param : bool {}
proc ref BaseArr.doiCanBulkTransferFrom(const ref Src: [], const viewDom: domain) param : bool {}
```

Since these are param methods, they must be resolved at compile-time. That means that they can't be virtual overloads of BaseArr versions. So, technically, we should document in/near BaseArr that we want this interface, but shouldn't create these methods in BaseArr since they should never be called there. The code in ChapelArray will be able to use them because there the full type is known. Generally code in ChapelDistribution will not have access to these methods.

**Comment by benharsh** [comment]
@mppf : good point. I just updated the text to indicate that these should be methods on classes inheriting ``BaseArr``.

**Comment by benharsh** [comment]
I just rewrote this issue with some different options and new information based on discussions with @vasslitvinov . I'm hoping to push forward and close this issue in the next couple of weeks.

**Comment by gbtitus** [comment]
This is something of a nit-pick, but with regard to:

> In these proposals I also recommend that the functions attempting to bulk-transfer return a bool indicating whether the bulk-transfer was successful. If not, the caller will be responsible for completing the array assignment.

By "completing" in that 2nd sentence we actually mean "performing", right?  At least to me, "completing" implies that the bulk-transfer function might have done part of the array assignment and the caller needs to do the rest.  But this isn't practical because the returned information doesn't have any way to communicate what was or wasn't done.  The function either has to do the entire assignment and return `true` or transfer nothing and return `false`.

**Comment by benharsh** [comment]
That's a good point. I intended to say that the caller should **perform** the bulk-transfer. I'll update the text.

**Comment by benharsh** [comment]
After developing these paths further, I've run into some technical and level-of-effort issues with the proposed interfaces where generics are concerned. Closing for now until I have some answers.

**Comment by benharsh** [comment]
I believe I have found a solution that handles the generic issues I ran into before. Instead of three proposals, I now have just one. This proposal has been implemented and passes testing.

**Comment by benharsh** [comment]
This interface was implemented with #7775.
